In [312]:
#reading dataset
#importing Dataset
import numpy as np
import pandas as pd
import random
from sklearn.datasets import load_iris
from scipy.spatial import distance
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['class'] = iris.target
df['class.names'] = [iris.target_names[i] for i in iris.target]
df.drop('class', axis=1, inplace=True)
df.head()
df = df.drop("class.names",1)

In [313]:
import random
num_clust = 4
centroid = []
num_features = 4
for i in range(0, num_clust):
    cent =[]
    for column in df:
        cent.append(random.randint(round(df[column].values.min()),round(df[column].values.max())))
    centroid.append(cent)
centroid_variation = [1,2,3,4]
p = 0
while np.mean(centroid_variation) > 0.001:
    print("#iteration")
    print(p+1)
    new_centroid = []
    centroid_variation = []
    clusters = []
    for index,row in df.iterrows():
        coordinates = []
        distances = []
        for column in range(0, num_features):
            coordinates.append(df.iloc[index,column])
        #print(coordinates)
        #print(centroid)
        for k in centroid:
            distances.append(distance.euclidean(coordinates,k))
        #print(distances)
        df.loc[index,"cluster"] = distances.index(min(distances)) + 1
        #print(df.loc[index,"cluster"])
    print("cluster - levels")
    df.cluster = df.cluster.astype(int)
    print(df["cluster"].unique())
    if len(df["cluster"].unique()) < num_clust:
        df = df.drop("cluster",1)
        for i in range(0, num_clust):
            cent =[]
            for column in df:
                cent.append(random.randint(round(df[column].values.min()),round(df[column].values.max())))
            centroid.append(cent)
        continue
    for i in range(1,num_clust+1):
        print(i)
        l = df.loc[df["cluster"]== i]
        n = l.shape[0]
        print(l.shape)
        m = []
        for column in l:
            m.append(l[column].mean())
        #print(m)
        new_centroid.append(m[0:4])
    print(new_centroid)
    for i in range(0, num_clust):
        #print(new_centroid[i])
        #print(centroid[i])
        centroid_variation.append(distance.euclidean(new_centroid[i],centroid[i]))
    #print("centroid variation")
    print(centroid_variation)
    centroid = new_centroid
    p = p+1
 

#iteration
1
cluster - levels
[3 1 2 4]
1
(4, 5)
2
(69, 5)
3
(73, 5)
4
(4, 5)
[[6.15, 3.6, 2.85, 0.825], [6.552173913043475, 2.9637681159420293, 5.308695652173915, 1.8623188405797098], [5.182191780821919, 3.121917808219178, 2.2972602739726025, 0.5657534246575342], [5.375, 2.825, 4.6, 1.675]]
[1.7278961195627471, 1.9774511269465769, 1.4810417638688953, 2.3101677428273475]
#iteration
2
cluster - levels
[3 2 4 1]
1
(1, 5)
2
(56, 5)
3
(53, 5)
4
(40, 5)
[[5.7, 2.6, 3.5, 1.0], [6.698214285714282, 3.0178571428571423, 5.437499999999999, 1.9249999999999996], [5.005660377358491, 3.360377358490567, 1.562264150943396, 0.2886792452830188], [5.76, 2.71, 4.325000000000001, 1.3924999999999998]]
[1.2867109232457772, 0.21159607994390275, 0.8396521275316499, 0.5629220638774067]
#iteration
3
cluster - levels
[3 2 4 1]
1
(10, 5)
2
(47, 5)
3
(50, 5)
4
(43, 5)
[[5.390000000000001, 2.42, 3.56, 1.06], [6.7553191489361675, 3.0361702127659567, 5.585106382978722, 2.004255319148936], [5.005999999999999, 3.41800000

In [ ]:
#points to ponder
#1. the number of clusters for some reason reduces to 3 sometimes to randomization

In [ ]:
    #clusters = []
    num_features = 4
    for index,row in df.iterrows():
        coordinates = []
        distances = []
        for column in range(0, num_features):
            coordinates.append(df.iloc[index,column])
        print(coordinates)
        for k in centroid:
            distances.append(distance.euclidean(coordinates,k))
        print(distances)
        df.loc[index,"cluster"] = distances.index(min(distances)) + 1
        




In [257]:
print(df["cluster"].unique())
centroid_variation

[2 3 1]


[0.0, 0.0, 0.0]

In [195]:
centroid = []
for i in range(1, 5):
    cent =[]
    for column in df:
        cent.append(random.randint(round(df[column].values.min()),round(df[column].values.max())))
    centroid.append(cent)

In [296]:
len(df["cluster"].unique())

4

In [ ]:
coordinates

In [237]:
centroid

[[5, 4, 5, 0], [7, 2, 3, 1], [7, 4, 1, 1], [4, 2, 1, 0]]

In [144]:
centroid.index(min(centroid))

3

In [ ]:
k = pd.DataFrame([1,2,3,4])
print(df)

In [132]:
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [ ]:
 for index,row in df.iterrows():
    coordinates = []
    distances = []
    print(index)
    print(row)
    print(df.iloc[index,:])
    #for column in range(0, df.shape[1]-1):
            #row.iloc[1,column]

#iteration
1
cluster - levels
[2 4 1 3]
(19, 5)
[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685, 1.0]
[[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685]]
(55, 5)
[5.029090909090909, 3.3381818181818192, 1.6345454545454543, 0.31999999999999995, 2.0]
[[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685], [5.029090909090909, 3.3381818181818192, 1.6345454545454543, 0.31999999999999995]]
(59, 5)
[6.545762711864404, 2.944067796610169, 5.428813559322033, 1.9406779661016944, 3.0]
[[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685], [5.029090909090909, 3.3381818181818192, 1.6345454545454543, 0.31999999999999995], [6.545762711864404, 2.944067796610169, 5.428813559322033, 1.9406779661016944]]
[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685]
[4, 2, 6, 0]
centroid variation
[2.810260188464728]
[5.029090909090909, 3.3381818181818192, 1.6345454545454543, 0.31999999999999995]
[6, 4, 1, 1]
centroid variation
[2.810260188464728, 1.498571220356022]
[6.545762711864404, 2.944067796610169, 5.428813559322033, 1.9406779661016944]
[7, 3, 7, 1]
centroid variation
[2.810260188464728, 1.498571220356022, 1.8875809366987]
#iteration
2
cluster - levels
[2 3 1]
(37, 5)
[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677, 1.0]
[[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(63, 5)
[6.601587301587299, 2.9857142857142858, 5.384126984126984, 1.9158730158730155, 3.0]
[[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.601587301587299, 2.9857142857142858, 5.384126984126984, 1.9158730158730155]]
[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677]
[5.531578947368421, 2.610526315789474, 4.1, 1.2526315789473685]
centroid variation
[0.1675083275510539]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.029090909090909, 3.3381818181818192, 1.6345454545454543, 0.31999999999999995]
centroid variation
[0.1675083275510539, 0.2043670332943172]
[6.601587301587299, 2.9857142857142858, 5.384126984126984, 1.9158730158730155]
[6.545762711864404, 2.944067796610169, 5.428813559322033, 1.9406779661016944]
centroid variation
[0.1675083275510539, 0.2043670332943172, 0.0863886012789818]
#iteration
3
cluster - levels
[2 3 1]
(41, 5)
[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998, 1.0]
[[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(59, 5)
[6.632203389830504, 2.9983050847457617, 5.4305084745762695, 1.9372881355932199, 3.0]
[[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.632203389830504, 2.9983050847457617, 5.4305084745762695, 1.9372881355932199]]
[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998]
[5.683783783783783, 2.678378378378378, 4.091891891891892, 1.2675675675675677]
centroid variation
[0.08234840869961255]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.08234840869961255, 0.0]
[6.632203389830504, 2.9983050847457617, 5.4305084745762695, 1.9372881355932199]
[6.601587301587299, 2.9857142857142858, 5.384126984126984, 1.9158730158730155]
centroid variation
[0.08234840869961255, 0.0, 0.06087465062318357]
#iteration
4
cluster - levels
[2 3 1]
(46, 5)
[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739, 1.0]
[[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(54, 5)
[6.664814814814812, 3.0074074074074066, 5.499999999999998, 1.9685185185185186, 3.0]
[[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.664814814814812, 3.0074074074074066, 5.499999999999998, 1.9685185185185186]]
[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739]
[5.729268292682925, 2.690243902439024, 4.151219512195122, 1.2999999999999998]
centroid variation
[0.09203331143350423]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.09203331143350423, 0.0]
[6.664814814814812, 3.0074074074074066, 5.499999999999998, 1.9685185185185186]
[6.632203389830504, 2.9983050847457617, 5.4305084745762695, 1.9372881355932199]
centroid variation
[0.09203331143350423, 0.0, 0.08337125548245827]
#iteration
5
cluster - levels
[2 3 1]
(50, 5)
[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36, 1.0]
[[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(50, 5)
[6.701999999999997, 3.015999999999999, 5.555999999999998, 1.992, 3.0]
[[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.701999999999997, 3.015999999999999, 5.555999999999998, 1.992]]
[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36]
[5.789130434782607, 2.71304347826087, 4.208695652173914, 1.3326086956521739]
centroid variation
[0.06551404992244121]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.06551404992244121, 0.0]
[6.701999999999997, 3.015999999999999, 5.555999999999998, 1.992]
[6.664814814814812, 3.0074074074074066, 5.499999999999998, 1.9685185185185186]
centroid variation
[0.06551404992244121, 0.0, 0.07172134004105034]
#iteration
6
cluster - levels
[2 3 1]
(54, 5)
[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926, 1.0]
[[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(46, 5)
[6.769565217391302, 3.0369565217391297, 5.599999999999998, 2.008695652173913, 3.0]
[[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.769565217391302, 3.0369565217391297, 5.599999999999998, 2.008695652173913]]
[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926]
[5.821999999999998, 2.7280000000000006, 4.256000000000001, 1.36]
centroid variation
[0.06776275891076816]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.06776275891076816, 0.0]
[6.769565217391302, 3.0369565217391297, 5.599999999999998, 2.008695652173913]
[6.701999999999997, 3.015999999999999, 5.555999999999998, 1.992]
centroid variation
[0.06776275891076816, 0.0, 0.08496457618354376]
#iteration
7
cluster - levels
[2 3 1]
(57, 5)
[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613, 1.0]
[[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(43, 5)
[6.802325581395348, 3.0441860465116273, 5.648837209302324, 2.0302325581395344, 3.0]
[[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.802325581395348, 3.0441860465116273, 5.648837209302324, 2.0302325581395344]]
[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613]
[5.829629629629629, 2.7314814814814814, 4.314814814814816, 1.3925925925925926]
centroid variation
[0.0440011794889611]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.0440011794889611, 0.0]
[6.802325581395348, 3.0441860465116273, 5.648837209302324, 2.0302325581395344]
[6.769565217391302, 3.0369565217391297, 5.599999999999998, 2.008695652173913]
centroid variation
[0.0440011794889611, 0.0, 0.06304299175265204]
#iteration
8
cluster - levels
[2 3 1]
(60, 5)
[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332, 1.0]
[[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(40, 5)
[6.827499999999999, 3.0699999999999994, 5.699999999999998, 2.062499999999999, 3.0]
[[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.827499999999999, 3.0699999999999994, 5.699999999999998, 2.062499999999999]]
[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332]
[5.8543859649122805, 2.7421052631578946, 4.34561403508772, 1.4087719298245613]
centroid variation
[0.04469158357399675]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.04469158357399675, 0.0]
[6.827499999999999, 3.0699999999999994, 5.699999999999998, 2.062499999999999]
[6.802325581395348, 3.0441860465116273, 5.648837209302324, 2.0302325581395344]
centroid variation
[0.04469158357399675, 0.0, 0.07041967411863342]
#iteration
9
cluster - levels
[2 3 1]
(61, 5)
[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966, 1.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(39, 5)
[6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153, 3.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153]]
[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966]
[5.885000000000001, 2.74, 4.376666666666667, 1.4183333333333332]
centroid variation
[0.020062423113003217]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.020062423113003217, 0.0]
[6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153]
[6.827499999999999, 3.0699999999999994, 5.699999999999998, 2.062499999999999]
centroid variation
[0.020062423113003217, 0.0, 0.03245957886903422]
#iteration
10
cluster - levels
[2 3 1]
(61, 5)
[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966, 1.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966]]
(50, 5)
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999, 2.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]]
(39, 5)
[6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153, 3.0]
[[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966], [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999], [6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153]]
[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966]
[5.88360655737705, 2.740983606557377, 4.388524590163935, 1.4344262295081966]
centroid variation
[0.0]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
[5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
centroid variation
[0.0, 0.0]
[6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153]
[6.853846153846153, 3.0769230769230766, 5.715384615384615, 2.053846153846153]
centroid variation
[0.0, 0.0, 0.0]